File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
class Solution {
    public String solution(String m, String[] musicinfos) {
        String answerTitle = "(None)";
        int answerPlayTime = 0;
        
        m = changeLower(m);
        for (String musicinfo : musicinfos) {
            String[] parts = musicinfo.split(",");
            String sheet = changeLower(parts[3]);
            int playTime = calculatePlayTime(parts[0], parts[1]);
            
            int i = 0;
            for (int j = 0; j < playTime; j++) {
                if (m.charAt(i) == sheet.charAt(j % sheet.length())) {
                    i++;
                    if (i == m.length()) {
                        if (answerPlayTime < playTime) {
                            answerTitle = parts[2];
                            answerPlayTime = playTime;
                        }
                        break;
                    }
                } else if (m.charAt(0) == sheet.charAt(j % sheet.length())) {
                    i = 1;
                    if (i == m.length()) {
                        if (answerPlayTime < playTime) {
                            answerTitle = parts[2];
                            answerPlayTime = playTime;
                        }
                        break;
                    }
                } else {
                    i = 0;
                }
            }
        }
        return answerTitle;
    }
    
    private static String changeLower(String sheet) {
        return sheet.replace("C#", "c")
                    .replace("D#", "d")
                    .replace("F#", "f")
                    .replace("G#", "g")
                    .replace("A#", "a")
                    .replace("B#", "b");
    }
    
    private static int calculatePlayTime(String start, String end) {
        String[] startParts = start.split(":");
        String[] endParts = end.split(":");
        return (Integer.parseInt(endParts[0]) * 60 + Integer.parseInt(endParts[1])) - (Integer.parseInt(startParts[0]) * 60 + Integer.parseInt(startParts[1]));
    }
}

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def calcultate_playingtime(start: str, end: str) -> int:
    start_h, start_m = map(int, start.split(':'))
    end_h, end_m = map(int, end.split(':'))
    return (end_h * 60 + end_m) - (start_h * 60 + start_m)



def solution(m: str, musicinfos: list) -> str:
    answer = ['', 0]
    
    m = m.replace('C#', 'c').replace('D#', 'd').replace('F#', 'f').replace('G#', 'g').replace('A#', 'a').replace('B#', 'b')
    for musicinfo in musicinfos:
        start, end, title, sheet = musicinfo.split(',')
        sheet = sheet.replace('C#', 'c').replace('D#', 'd').replace('F#', 'f').replace('G#', 'g').replace('A#', 'a').replace('B#', 'b')
        playingtime = calcultate_playingtime(start, end)
        i, len_sheet, len_m = 0, len(sheet), len(m)
        for j in range(playingtime):
            if m[i] == sheet[j % len_sheet]:
                i += 1
                if i == len_m:
                    if answer[1] < playingtime:
                        answer = [title, playingtime]
                    break
            elif m[0] == sheet[j % len_sheet]:
                i = 1
                if i == len_m:
                    if answer[1] < playingtime:
                        answer = [title, playingtime]
                    break
            else:
                i = 0
    
    return answer[0] if answer[0] != '' else '(None)'